### PR TITLE
Solving the validation hanging issue

### PR DESCRIPTION
**Pull request by DNXie**
As the [discussion](https://github.com/pytorch/torchtitan/issues/1618#issuecomment-3215472411), I added:
- warning message when the validation `steps`=-1 in both comment and logger
- change the default `steps` to reasonable values with the common setup (world size = 8).  
- Add infinite loop support for validator to avoid hang when `steps` is large enough to exhaust the dataset.
- Add the same fix for flux.

## Test
- 8 GPUs with `steps=-1`: hang around step 1270
- 8 GPUs with `steps=1200`: good
- 8 GPUs with `steps=1500`: `infinite` automatically set to true. Exhaust the dataset and re-iterate, but won't hang
- Flux: `steps=-1` doesn't hang
- Flux: `steps=60` doesn't hang; re-loop the dataset.

Full thread: https://github.com/pytorch/torchtitan/issues/1618

cc @ebsmothers @tianyu-l

### CODE REVIEW

File: torchtitan/models/llama3/train_configs/llama3_70b.toml [hunk 1]
```diff
@@ -67,4 +67,6 @@ filter_fqns = ["output"]
 enabled = false
 dataset = "c4_validation"
 freq = 500
-steps = -1
+# Set steps to -1 to consume all the data in the dataset
+# WARNING: When setting to -1 there could be hangs due to mismatch among ranks
```
Comment (fegin) [inline torchtitan/models/llama3/train_configs/llama3_70b.toml]: Instead of adding the comment to every toml, can't we add the comment to `job_config.py`? And since you also add a runtime warning, I think it should be enough.

File: torchtitan/models/llama3/train_configs/llama3_70b.toml [hunk 2]
```diff
@@ -67,4 +67,6 @@ filter_fqns = ["output"]
...
+steps = 1200 # Recommend value for c4_validation with world-size = 8
```
Comment (wwwjn) [inline torchtitan/models/llama3/train_configs/llama3_70b.toml]: I feel like it's a little bit confusing to set 1200 in this toml file - training.step = 1000, but validation step=1200 in the toml file? How did we handle if validation.step > trainining.step

Comment (wwwjn) [reply torchtitan/models/llama3/train_configs/llama3_70b.toml]: And also validation.freq=500, so is it better to set validation.steps=1000?

File: torchtitan/components/validate.py [hunk 3]
```diff
@@ -67,6 +68,7 @@ def __init__(
             dp_world_size=dp_world_size,
             dp_rank=dp_rank,
             tokenizer=tokenizer,
+            infinite=self.job_config.validation.steps != -1,
```
Comment (wwwjn) [inline torchtitan/components/validate.py:70]: I think we need to at least let user realize the validation dataset if going to be re-looped, if the dataloader run out of data. I don't think infinite should be set based on steps silently.

File: torchtitan/models/llama3/train_configs/llama3_70b.toml (hunk 2 above)
Comment (tianyu-l) [reply torchtitan/models/llama3/train_configs/llama3_70b.toml]: > How did we handle if validation.step > trainining.step

They are independent. `validation.step` means how many steps of validation we run on the validation dataset in a single validation step.

Comment (DNXie) [reply torchtitan/models/llama3/train_configs/llama3_70b.toml]: `freq` is how frequent we run the validation. If `freq=500`, it means we run validation every 500 training steps.

File: torchtitan/components/validate.py (hunk 3 above)
Comment (DNXie) [reply torchtitan/components/validate.py:70]: There will be a warning if we are reiterating [here](https://github.com/pytorch/torchtitan/blob/030879fa59be48e3edafe997a00877671c461e6c/torchtitan/datasets/hf_datasets.py#L146)
Do you think it is enough? 
This `infinite` is mainly to avoid the hang is user set `steps` too large that exhausts the dataset.
See test number 3, if steps=1500, without this `infinite=False`, the validation would hang.

File: torchtitan/models/llama3/train_configs/llama3_70b.toml (hunk 1 above)
Comment (DNXie) [reply torchtitan/models/llama3/train_configs/llama3_70b.toml]: Good point. I have removed the warning comments in the toml files and added a warning comment in `job_config.py`

File: torchtitan/models/llama3/train_configs/llama3_70b.toml (hunk 2 above)
Comment (wwwjn) [reply torchtitan/models/llama3/train_configs/llama3_70b.toml]: Gotcha, thanks for explain!

File: torchtitan/components/validate.py (hunk 3 above)
Comment (wwwjn) [reply torchtitan/components/validate.py:70]: I see, the warning is enough. I thought about this setting here again and I think the setting makes sense to me.

For validation, I think a common use case is:  user wants to calculate loss over all the validation dataset, so the user would use step=-1 for now. Since we can not get the dataset size before iterating. If the user wants to consume  `steps * local_batch_size * dp_rank` samples during validation, we should follow their settings faithfully.

File: torchtitan/models/llama3/train_configs/llama3_405b.toml [hunk 4]
```diff
@@ -68,4 +68,4 @@ filter_fqns = ["output"]
 enable = false
 dataset = "c4_validation"
 freq = 500
-steps = -1
+steps = 1200  # Recommend value for c4_validation with world-size = 8
```
Comment (tianyu-l) [inline torchtitan/models/llama3/train_configs/llama3_405b.toml]: This is not only related to world size, but also sequence length, as we use sequence packing for data loading.

File: torchtitan/components/validate.py (hunk 3 above)
Comment (DNXie) [reply torchtitan/components/validate.py:70]: Yes, when `steps=-1`, it would set `infinite` to False, which is faithful to what user intend to do. But this almost always lead to a hang. 

It would only set `infinite` to True if `steps!=-1`. 

Do you mean the current code is good?

Comment (wwwjn) [reply torchtitan/components/validate.py:70]: Yes! This part looks good to me, thanks for discussing!

File: torchtitan/models/llama3/train_configs/llama3_405b.toml (hunk 4 above)
Comment (DNXie) [reply torchtitan/models/llama3/train_configs/llama3_405b.toml]: Updated. Please check

**Comment by DNXie** [comment]
> Please address in flux/validation.py as well: https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/validate.py

@wwwjn I have added the fix to flux scripts. Please check. I didn't update the `steps` in toml files because the current setting [`steps=1`](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/train_configs/debug_model.toml#L72)  doesn't cause hang.

**Comment by wwwjn** [comment]
> > Please address in flux/validation.py as well: https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/validate.py
> 
> @wwwjn I have added the fix to flux scripts. Please check. I didn't update the `steps` in toml files because the current setting [`steps=1`](https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/flux/train_configs/debug_model.toml#L72) doesn't cause hang.

This part looks good to me! I will let @wesleytruong take a look as well

**Comment by wesleytruong** [comment]
Thanks for adding these changes! I was planning to look into this validation dataloader issue so this helps a lot as a workaround.

I think we still need to specify a default steps =1 here https://github.com/pytorch/torchtitan/blob/46a32e7883e8ab22b91007b5f5c77d9326e35480/torchtitan/experiments/flux/train_configs/flux_schnell_model.toml#L72
Currently steps isn't provided so it defaults to -1. 
It would be even better if you could provide a good default steps instead of 1 in `flux_dev_model.toml` and `flux_schnell_model.toml`. I looked at the coco validation set and the validation set is around ~3k samples. You may need to account for the different `local_batch_size` in dev and schnell models, and also the potential images that may be smaller than 256x256 and getting dropped. Looking at the dataset this should only be around <~1%.

If it's not too much work this would be great, but otherwise changes look good to me!

File: torchtitan/components/validate.py [hunk 5]
```diff
@@ -74,6 +76,11 @@ def __init__(
         self.pp_has_first_stage = pp_has_first_stage
         self.pp_has_last_stage = pp_has_last_stage
 
+        if self.job_config.validation.steps == -1:
+            logger.warning(
+                "Setting validation steps to -1 could cause hangs due to mismatch among ranks."
```
Comment (wwwjn) [inline torchtitan/components/validate.py:81]:
```suggestion
                "Setting validation steps to -1 might cause hangs because of unequal sample counts across ranks when dataset is exhausted"
```

Please change other places and make the warning more descriptive before land

**Comment by DNXie** [comment]
> Thanks for adding these changes! I was planning to look into this validation dataloader issue so this helps a lot as a workaround.
> 
> I think we still need to specify a default steps =1 here
> 
> https://github.com/pytorch/torchtitan/blob/46a32e7883e8ab22b91007b5f5c77d9326e35480/torchtitan/experiments/flux/train_configs/flux_schnell_model.toml#L72
> 
> 
> Currently steps isn't provided so it defaults to -1.
> It would be even better if you could provide a good default steps instead of 1 in `flux_dev_model.toml` and `flux_schnell_model.toml`. I looked at the coco validation set and the validation set is around ~3k samples. You may need to account for the different `local_batch_size` in dev and schnell models, and also the potential images that may be smaller than 256x256 and getting dropped. Looking at the dataset this should only be around <~1%.
> If it's not too much work this would be great, but otherwise changes look good to me!

@wesleytruong Just updated the default `steps` in all flux configs. For flux experiments, the validation doesn't hang even with `steps=-1`, so I set the steps to exact number to exhaust the validation dataset. I guess this is because the number of samples (after packing dropping low quality ones) can be divided by `batch_size*world_size` always.
